### ISSUE TITLE
CPU usage slowly increases over time

### ISSUE THREAD

**Issue by stevedesmond-ca** [Microsoft/ApplicationInsights-aspnetcore#690]
I had previously used Application Insights, but no longer do, however the App Insights code remained in my app. I noticed when I recently migrated hosts and upgraded to .NET Core 2.1.0-rc1 that the CPU usage gradually increased over the course of a few days.
![image](https://user-images.githubusercontent.com/195896/40263271-d32fbaa2-5add-11e8-9c70-bb3f9f8677bb.png)
The drops back to 0 are Docker container restarts.

Removing Application Insights from the site results in a steady almost-zero CPU usage.

See also [this Twitter thread](https://twitter.com/stevedesmond_ca/status/996919794589995009), though there's not much more there.

## Repro Steps
1. Add Application Insights (2.3.0-beta2) to an ASP.NET Core MVC app via NuGet and Startup.cs
`app.AddApplicationInsightsTelemetry("{no-longer-valid instrumentation key}")`
1. Setup some sort of synthetic monitoring to hit the site every minute
1. Wait a couple days ¯\_(ツ)_/¯

##  Actual Behavior
- CPU slowly creeps up over the course of a few days

## Expected Behavior
- CPU remains stable

## Version Info 
SDK Version : 2.1.0-rc1
.NET Version : 2.1.0-rc1
How Application was onboarded with SDK(VisualStudio/StatusMonitor/Azure Extension) : 
OS : Ubuntu 16.04
Hosting Info (IIS/Azure WebApps/ etc) : Azure VM

**Comment by cijothomas** [comment]
Thanks for reporting.  We'll try to setup a repro and investigate more.
What was the version of application insights SDK which you used?

**Comment by stevedesmond-ca** [comment]
2.3.0-beta2, running on an ASP.NET Core 2.1-rc1 app.

**Comment by cijothomas** [comment]
![image](https://user-images.githubusercontent.com/5232798/40672587-1203ba6e-6324-11e8-88bb-f708154e3d3f.png)
Left an app in .net 2.0 running for last 3 days. Dont see any increase in mem/cpu

**Comment by cijothomas** [comment]
![image](https://user-images.githubusercontent.com/5232798/40672643-3ee86520-6324-11e8-9f7f-7861532a761b.png)
Similar for app in .net 2.1 for last 3 days.

**Comment by cijothomas** [comment]
I'll setup something for linux now.

**Comment by thohng** [comment]
![image](https://user-images.githubusercontent.com/9668549/40698003-84f42410-63ff-11e8-930a-2e0eb2e09a96.png)

We got same issue on Ubuntu 16.04
- After upgrade AspNetCore 2.1.0-rc1
- Never got CPU issue with AspNetCore 2.0.x before
- Not sure this caused by new AspNetCore 2.1.0-rc1 or new ApplicationInsights 2.3.0-beta2
- Internally, we got few HttpClient failed a day, might related to https://github.com/dotnet/corefx/issues/29285

Before Upgrade:
```
<PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.2.1" />
<PackageReference Include="Microsoft.AspNetCore.ApplicationInsights.HostingStartup" Version="2.0.2" />
<PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.0.3" />
```

After:
```
<PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.3.0-beta2" />
<PackageReference Include="Microsoft.AspNetCore.ApplicationInsights.HostingStartup" Version="2.1.0-rc1-final" />
<PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.1.0-rc1-final" />
```

**Comment by cijothomas** [comment]
Thanks for reporting this. We will setup a linux test to do deeper investigation soon.

**Comment by cijothomas** [comment]
Have an asp.net core 2.1 app with AI SDK 2.3, running in ubuntu 16_04. A script is pinging the app once every 10 sec. 
WIll let this run for few hours/day and check again.

**Comment by cijothomas** [comment]
6hours, no visible change in CPU load. CPU is constant around .39%
Load is 6Request/Sec
Will leave this overnight.

**Comment by thohng** [comment]
When we stop pinging, the CPU usage is no more keep constantly increase.

![image](https://user-images.githubusercontent.com/9668549/40818317-33ba791a-6589-11e8-87b5-2ba58812b167.png)

**Comment by mehmetilker** [comment]
Same thing happening on Windows as well (Azure App service)
Not only CPU but also Garbage collector affected.
I upgrade my .Net core 2.0 site to .Net core 2.1 on Jun 3 afternoon. And yesterday I have removed AI Insight extension from my app service (Nuget package reference still in my project.)
You can see changes in attached images.

![cpu](https://user-images.githubusercontent.com/279289/41141102-5758b43e-6af9-11e8-928e-b47e02625836.PNG)
![gen](https://user-images.githubusercontent.com/279289/41141103-577a82f8-6af9-11e8-931e-1071d7bf75c2.PNG)

**Comment by cijothomas** [comment]
repro-ed on linux vm. Need to investigate more.

**Comment by flagbug** [comment]
@stevedesmond-ca @mehmetilker @thohng I've seen a similar thing since upgrading to .NET Core 2.1 and not only does our CPU usage increase over time, but our open handles increase *dramatically* since the upgrade:

![capture](https://user-images.githubusercontent.com/964691/41307528-b9a4cd34-6e79-11e8-8b97-42c2252cfef8.PNG)

Could you take a look at your handle count and confirm if you have the same issue or not?

**Comment by cwe1ss** [comment]
I'm recently having issues with increased memory usage. I checked it with dotMemory and I've seen many System.Net.Http related objects - e.g. open handlers that haven't been closed etc.

I'm not yet sure if it's Application Insights though (I will do some more tests next week) - but if it is, maybe this clue helps finding the issue.

**Comment by ByronAP** [comment]
I also have increasing CPU which seems to be from HTTP. It was not noticeable until I went to production which looked great after deployment, but a few hours later everything was redlined due to ever-increasing CPU demand as shown below. Unfortunately, this is a show stopper and is forcing me to rollback.

Runtime: core 2.1 Linux x64
Containers: YES Docker 
OS: UBUNTU & CENTOS

![cpu](https://user-images.githubusercontent.com/7310089/41560342-c9f3b20a-7314-11e8-92e0-56c77f98585d.PNG)

**Comment by flagbug** [comment]
We've downgraded Application Insights back to v2.2.1 and don't see the increasing CPU and memory usage anymore, so I can confirm that this is an issue with Application Insights.

@cijothomas Any chance this can be looked into? This basically makes version 2.3.0 not usable for production scenarios

**Comment by cijothomas** [comment]
@flagbug Thanks for helping narrow down to 2.3.0.  We will prioritize this investigation.

**Comment by cijothomas** [comment]
Related, maybe the cause as well: https://github.com/Microsoft/ApplicationInsights-dotnet/issues/594

**Comment by nycdotnet** [comment]
Might want to see if `HttpClientFactory` solves the issue.  Static `HttpClient` is not without its issues. https://blogs.msdn.microsoft.com/webdev/2018/02/28/asp-net-core-2-1-preview1-introducing-httpclient-factory/

**Comment by cwe1ss** [comment]
I now also ran our application without App Insights for a few days and I don't see any stale Http-related objects in memory so I'm also "blaming" Application Insights for now.

Could you please prioritize this? This has been open for more than a month and a profiler that leaks memory is a "really bad thing" ™️...

/cc @SergeyKanzhelev @lmolkova (fyi, in case you didn't know about this yet)

**Comment by cijothomas** [comment]
@cwe1ss Were you facing issues in linux or windows or both?

**Comment by cwe1ss** [comment]
My setup: Windows Server 2016 on Azure (Standard_D11_v2, 2 cores 16GB RAM); Single-node Service Fabric cluster (development system); I'm running ~ 60 .NET Core apps (~ 30 ASP.NET Core apps, 30 console apps); services are communicating via the Service Fabric HTTP reverse proxy. As it's a dev system, there's very little load on the system.

Here's some more details:

I see a similar pattern in overall CPU usage like the other reporters:
![cpu-increase](https://user-images.githubusercontent.com/4581460/42150598-a84bf178-7dda-11e8-8554-d9c7d1e38103.png)

I took a process dump of one ASP.NET core Web API project. The process has been running for "2 days 20:49:36". dotMemory gives me the following overview. As you can see, the string value "dc.services.visualstudio.com" exists 23218 times.
![dotmemory-overview](https://user-images.githubusercontent.com/4581460/42150824-45075840-7ddb-11e8-9de1-d737f00287cc.png)

Looking at the type list shows that "System.Net.Http"-related objects make up the top of the list.
![types-list](https://user-images.githubusercontent.com/4581460/42150892-7d7d099a-7ddb-11e8-8322-54508147e264.png)

I do not see a similar pattern in an app where I've disabled Application Insights.

**Comment by benaadams** [comment]
@cwe1ss what are the back traces/call tree like on some of these objects?

11,613 object count ones are interesting; especially since it looks like there are 11,613 sets of `HttpConnectionPoolManagers` and 11,613 x `List<HttpConnectionPool+CachedConnection>`. That's not the number of cached connections; but the number of caches for connections.

`HttpClientHandler` probably would have the most informative backtraces? (as I assume all the others are allocated in it)

**Comment by cwe1ss** [comment]
@benaadams can I get this from the process dump via dotMemory / WinDbg or do I need to trace this with PerfView / dotTrace ?

**Comment by cwe1ss** [comment]
oh... I've seen the tabs in dotMemory... but they are empty and just say "No allocation info was collected during profiling". So I guess I need PerfView / dotTrace, right?!

**Comment by cwe1ss** [comment]
I clicked around a bit more in dotMemory (I don't know much about it, sorry) and I've seen that you can get more details about the object instances. Here's some screenshots for a random instance of `System.Collections.Concurrent.ConcurrentDictionary+Tables<System.Net.Http.HttpConnectionPoolManager+HttpConnectionKey, System.Net.Http.HttpConnectionPool>`.

![dotmemory-instance](https://user-images.githubusercontent.com/4581460/42160917-3ac30488-7df9-11e8-9d95-2fb47c2d4564.png)

![dotmemory-instance-shortestpath](https://user-images.githubusercontent.com/4581460/42160920-3d96b772-7df9-11e8-961f-0cf33eec2dd9.png)

![dotmemory-instance-path](https://user-images.githubusercontent.com/4581460/42160923-3f3d9956-7df9-11e8-943c-4e568d026c59.png)

Hope this helps?!

**Comment by benaadams** [comment]
I'm guessing the CPU increase is from the constantly increasing number of `Timer`s which you memory dump seems to indicate. `Timer`s also act as GC roots, which is why everything is probably hanging around, if their callbacks are capturing state.

Not sure why though...

**Comment by cwe1ss** [comment]
Could this be an issue with how [HeartbeatProvider uses `Timer`](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/a7ffc8e3f1aa0740baa53a7e1bfe167265b872f7/src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs#L161)? See e.g. https://www.codeproject.com/Answers/185740/Threading-Timer-prevents-GC-collection#answer1 where the guy from "Solution 1" says:

>The method must be static...
>Conversely, if you use a non static method and you reference it using this.NameMethod, the System.Threading.TimerCallback will keep a reference to each MyTimerData object. So, despite your mtd=null instruction, it won't be free until the closing of the applicaton.
>
>Your example is a circular reference:
mtd => m_timer => new System.Threading.TimerCallback(...) => mtd.TimerCall(o){}

@benaadams thank you very much for looking into this!!!

**Comment by benaadams** [comment]
> Could this be an issue with how HeartbeatProvider uses Timer?

Looks like its only created once though? Doesn't explain why there are 11,628 Timers in your memory capture.

**Comment by cwe1ss** [comment]
True. 😕 Well, I'll leave it to the Application Insights guys then. Or do you think this could be something that's wrong in my application?

Thanks again, Ben. I always learn something new when you're involved! 😉 👍 🍺

**Comment by benaadams** [comment]
My best guest on the second set of screenshots you provided is from `Task.Delay` with a `CancellationTokenSource`; but that might just be a low instance. (i.e. `Task.Delay(cts.Token)` tasks being created faster than they are completed)

Could you give from `Types` tab the lower pane, and open the call chain when you have Timer selected in the top panel? 

e.g.
![image](https://user-images.githubusercontent.com/1142958/42166300-7de736ae-7e02-11e8-8715-765355b2edcc.png)

That should hopefully highlight the call chain that is allocating them

**Comment by cwe1ss** [comment]
I currently only have a process dump (created from Task Manager) and I only get "No allocation info was collected during profiling" on the "Call Tree" tabs etc.

Do you know of a way to profile .NET Core apps on a Service Fabric cluster? Seems like Remote Attach doesn't yet work with .NET Core and I don't know of any way to start the profiling when Service Fabric is responsible for starting the program.

**Comment by benaadams** [comment]
> Do you know of a way to profile .NET Core apps on a Service Fabric cluster?

No, attach to running process doesn't work with DotTrace for a local process either.

However you could probably create a simple local app, add appInsights and profile that for similar effect?

**Comment by cwe1ss** [comment]
>No, attach to running process doesn't work with DotTrace for a local process either.

Too bad. I don't want to spend too much time on this so I'll wait for a response from the Application Insights team first.

**Comment by cwe1ss** [comment]
@cijothomas @SergeyKanzhelev another ping. Could you please bring this to the attention of someone?

I'm sorry to be annoying but Application Insights is a paid service and a potential memory leak is a very bad thing for an APM product, so some update would be appropriate IMO. It's now more than 6 weeks since this issue has been created and I feel like no one has properly looked into this yet.

**Comment by cijothomas** [comment]
Though I haven't understood things fully yet, here's my findings from today's experiments.
1. Ran Asp.Net Core 2.1 App with AI SDK 2.2. Dotmemory trace shows a bunch of  objects like `HttpConnectionPoolManager` `List<HttpConnectionPool+CachedConnection>` etc.. Their count kept increasing with time. (Similar to @cwe1ss 's snapshots.)
2. Ran Asp.Net Core 2.1 App with AI SDK 2.3. Same as # 1 

3. Ran Asp.Net Core 2.0 App with AI SDK 2.2. Don't see any `HttpConnectionPoolManager` in memory.
4. Ran Asp.Net Core 2.0 App with AI SDK 2.3. Don't see any `HttpConnectionPoolManager` in memory.

Though the issue of one-HttpClient-per-transmission existed for long time in AI SDK Channel (https://github.com/Microsoft/ApplicationInsights-dotnet/issues/594), the issue exasperated with Asp.Net Core 2.1. 

I validated this with a test app new-ing up HttpClient everytime it makes outbound call. It shows steady increase in `HttpConnectionPoolManager` counts in memory.

I'll make a fix to this issue: https://github.com/Microsoft/ApplicationInsights-dotnet/issues/594, and publish a beta SDK soon for others also to validate.

In my quick experiment with fixed SDK, the count of `HttpConnectionPoolManager` etc, is not increasing. (stays at 2 or 3 as expected).

**Comment by cijothomas** [comment]
Some folks mentioned that the issue occurs with AI SDK 2.3 but not with 2.2.1 is puzzling as there was no major change made to channel/transmission between these two versions. So there could be something else as well. 
But I'll tackle the known issue(Microsoft/ApplicationInsights-dotnet#594)) first anyways.

**Comment by cwe1ss** [comment]
@cijothomas thank you for looking into it! Let me know if there's anything else you'd like to see from my process dump!

**Comment by cwe1ss** [comment]
fyi: With .NET Core 2.1, the default for HttpClient changed to the new `SocketsHttpHandler` (https://blogs.msdn.microsoft.com/dotnet/2018/02/27/announcing-net-core-2-1-preview-1/). So if this issue happens due to the upgrade of .NET Core instead of the upgrade of the Application Insights SDK, maybe that's something to look at?!

**Comment by iwillspeak** [comment]
I _think_ we're seeing this issue with `Microsoft.ApplicationInsights.AspNetCore` v 2.2.1 on .NET Framework 4.7 (460805). CPU increases over time. Increase seems to be linked to the rate of HTTP requests being serviced by our app. Performance profiling shows that CPU is being spent waiting for the lock in `TimerQueue.FireNextTimers`, presumably because timers are being leaked.

**Comment by ericgreenmix** [comment]
We were seeing this on version 2.3.0 when we updated to .Net Core 2.1. Yesterday, I downgraded back to version 2.2.1 on .Net Core 2.1 after reading this issues thread. However, we are still seeing this exact same issue. So I don't think it is directly related to version 2.3.0.

**Comment by cijothomas** [comment]
Thanks all for the help so far- as I mentioned, this is not because of AI SDK 2.2.1 to 2.3 upgrade - AI SDK had incorrectly used HttpClient for a long time, but now with some underlying change in .net core 2.1, this is causing the leak.
The leaked `HttoConnectionPool` things are all rooted to some `Timer` object (can't recollect exactly).

I'll provide a fix today.

**Comment by thohng** [comment]
The issue is fixed after upgrade to new version `Microsoft.ApplicationInsights@2.7.0-beta3`

```
<PackageReference Include="Microsoft.ApplicationInsights" Version="2.7.0-beta3" />
<PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.4.0-beta2" />
<PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.2" />
<PackageReference Include="Microsoft.AspNetCore.ApplicationInsights.HostingStartup" Version="2.1.1" />
```

Thank you!

![image](https://user-images.githubusercontent.com/9668549/42741671-f0f425f0-88e6-11e8-82f0-c42f00f87d89.png)

**Comment by cijothomas** [comment]
@thohng Thanks for helping validate the fix in 2.7.0-beta3.
I am releasing Asp.net core sdk 2.4.0-beta3 today (which refers 2.7.0-beta3 of the base sdk containing the fix)

**Comment by cijothomas** [comment]
https://www.nuget.org/packages/Microsoft.ApplicationInsights.AspNetCore/2.4.0-beta3 - Please try this version and let us know if you still see issues.

**Comment by thohng** [comment]
Thanks @cijothomas , the `Microsoft.ApplicationInsights.AspNetCore@2.4.0-beta3` is good
```
<PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.4.0-beta3" />
<PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.2" />
<PackageReference Include="Microsoft.AspNetCore.ApplicationInsights.HostingStartup" Version="2.1.1" />
```
![image](https://user-images.githubusercontent.com/9668549/42808046-e9474824-89e4-11e8-8324-1c45fed17a47.png)

Compare with the graph before
![image](https://user-images.githubusercontent.com/9668549/42808139-26360c0c-89e5-11e8-9ecd-8043b0526fd3.png)

**Comment by vigneshmsft** [comment]
I have updated few of my aspnet core services to netcore21 and latest AppInsights SDK (2.1.1)
and the memory metrics in azure for my services are all creeping up every hour.
It used to be stable until before the update but is growing now.
I'll post more details later. Only change is to netcore 2.1 and app insights SDK (2.1.1)

**Comment by cijothomas** [comment]
@vigneshmsft can you upgrade you application insights sdk to the latest beta, and let us know if seeing this issue? (2.1.1 is not the latest, the latest stable is 2.3.0, and latest beta is 2.4.0-beta3)
https://www.nuget.org/packages/Microsoft.ApplicationInsights.AspNetCore

**Comment by vigneshmsft** [comment]
@cijothomas sorry I meant I am on 2.1.1 for the netcore sdk

Microsoft.AspNet.Core.App is 2.1.1
Microsoft.ApplicationInsights 2.6.4
Microsoft.ApplicationInsights.AspNetCore 2.3.0

And am not at liberty to update to beta version of any packages in production.

**Comment by cijothomas** [comment]
@vigneshmsft The fix for the incorrect HttpClient usage leading to leaking is available as a beta only today. We'll be releasing stable in next week most likely. I understand your concern about using beta in prod - if possible validate that this is the issue you are hitting by trying out the beta in non-prod environment. 

( The issue occurs only on on NetCore 2.1. 2.0 did not have this issue from our experiments. )

**Comment by vigneshmsft** [comment]
@cijothomas any updates on when a stable version will be released?

**Comment by cijothomas** [comment]
@vigneshmsft In ~ 1 week.

**Comment by cijothomas** [comment]
Most likely by Wednesday next week.

**Comment by Meenakshise** [comment]
Ours is a normal aspnet web api project. Please look at the package file. We too face the same CPU issue. Will this get resolved if i upgrade the application insights package to the  latest of 2.6.4 and 2.4.0 respectively. Please let  me know. This is not a .net core project but. This was created recently 

![image](https://user-images.githubusercontent.com/5516586/43708253-acf6c914-9987-11e8-9143-331bc593f0fb.png)

![image](https://user-images.githubusercontent.com/5516586/43708349-df89d970-9987-11e8-8e8a-0c9d5869df3a.png)

**Comment by cijothomas** [comment]
@Meenakshise The issue we addressed here was specific to .NET Core apps only, so i don't think this will help you. Either way, we always recommend using the latest version to get all the bug fixes etc.
Could you try to use the latest stable/beta, and report back if you have high CPU usage issue.

**Comment by vigneshmsft** [comment]
updating to 2.4.0 has fixed the memory leak issue.

**Comment by IanKemp** [comment]
Any explanation on what was the issue?

**Comment by cijothomas** [comment]
@IanKemp 
https://github.com/Microsoft/ApplicationInsights-dotnet/issues/594
https://github.com/Microsoft/ApplicationInsights-dotnet/pull/857

**Comment by Svisstack** [comment]
I upgraded to the `2.4.1` and it still has the leak which increases over time.

First screenshot showing 2 threads working only and constantly on the CPU leak, I found that the amount of threads depend on uptime of the process, so leak increasing in time and was not fixed in 2.4.1

Second screenshot showing the increase ratio where 600% is a hardware limit, after some time AspNetCore app working only on the leak.

@cijothomas Please advise here.

![1](https://user-images.githubusercontent.com/864295/45268207-dd7f9000-b478-11e8-8d8c-90e6e25fc7e8.PNG)

![2](https://user-images.githubusercontent.com/864295/45268240-40712700-b479-11e8-869a-bf88fc36d89a.PNG)

**Comment by benaadams** [comment]
@Svisstack what version of .NET are you running on? High CPU usage in the `HttpHeaders` `Dictionary'2.TryGetValue` would suggest concurrent writes to the header dictionary. 

.NET Core 2.1 will throw an exception when an infinite loop would previously occur due to this https://github.com/dotnet/corefx/issues/28123

**Comment by benaadams** [comment]
@Svisstack what version of .NET are you running on? High CPU usage in the `HttpHeaders` `Dictionary'2.TryGetValue` would suggest concurrent reads+writes to the header dictionary. 

.NET Core 2.1 will throw an exception when an infinite loop would previously occur due to this https://github.com/dotnet/corefx/issues/28123

It could happen if `FrameworkSqlEventListner` (or its chain in your callstack) if fired after/concurrently with the request end (when the Dictionary is cleared)

**Comment by Svisstack** [comment]
@benaadams I'm running this on the .NET Framework 4.7.2

If this is true it looks like a major bug in appinsights; why someone would put a SpinLock-like reader around the dictionary item which does not exist.

**Comment by benaadams** [comment]
>  why someone would put a SpinLock-like reader around the dictionary item which could not exist.

The dictionary for the header collection is reused for each request on the same connection; so at the end of the request it is cleared. If that clearing happens while another thread is reading the header collection (e.g. it reads it outside of the normal request flow); then the read can go into an infinite loop as the arrays inside the dictionary are changed (Dictionary's are not threadsafe for a concurrent write with any other operation on it). In .NET Core 2.1 it will detect the infinite loop being formed and throw an exception; though it won't fix the concurrent use.

**Comment by Svisstack** [comment]
@benaadams Thanks for detailed explanation. I understand now what's happening, but didn't see a clear solution of this as it looks like the bug is actually in the AspNetCore which allows concurrent requests to the dictionary which is not thread-safe.

**Comment by benaadams** [comment]
>  looks like the bug is actually in the AspNetCore which allows concurrent access to the dictionary instance

Well... even if concurrent access was prevented (at a performance cost); it still wouldn't help as the data it read would still be wrong. 

If its conflicting with `.Clear()` this means there is a race condition and 3 outcome could happen (if concurrent access was prevented)

1. It could read the right data (pre clear)
2. It could read no data (post clear) **wrong**
3. It could read headers from the next request **very wrong**

At a guess... Looking at the stack from dotTrace, it could be a Sql query that's started as part of the Request; but not `await`ed as part of the request, so the response is sent before the Sql is completed. It could also be a `Timer` that's started to periodically poll and lazily started as part of a Request.

When it completes it would then be trying to log the shallow copied (i.e. reference to) the captured Request data as part of the Application Insight's Telemetry event; using the mutable Request information that has now changed and is no longer valid for the event.

I'm not entirely sure how [`FrameworkSqlProcessing.OnEndExecuteCallback`](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/af20672f42145b837bf50b48e30e8427ca817e2d/Src/DependencyCollector/Shared/Implementation/FrameworkSqlProcessing.cs#L89-L110) accesses the Headers collection, though it looks to be part of the `DependencyTelemetry` object.
```csharp
public void OnEndExecuteCallback(long id, bool success, bool synchronous, int sqlExceptionNumber)
{
    DependencyCollectorEventSource.Log.EndCallbackCalled(id.ToString(CultureInfo.InvariantCulture));

    var telemetryTuple = this.TelemetryTable.Get(id);

    if (telemetryTuple == null)
    {
        DependencyCollectorEventSource.Log.EndCallbackWithNoBegin(id.ToString(CultureInfo.InvariantCulture));
        return;
    }

    if (!telemetryTuple.Item2)
    {
        this.TelemetryTable.Remove(id);
        var telemetry = telemetryTuple.Item1 as DependencyTelemetry;
        telemetry.Success = success;
        telemetry.ResultCode = sqlExceptionNumber != 0 ? sqlExceptionNumber.ToString(CultureInfo.InvariantCulture) : string.Empty;
        DependencyCollectorEventSource.Log.AutoTrackingDependencyItem(telemetry.Name);
        ClientServerDependencyTracker.EndTracking(this.telemetryClient, telemetry);
    }
}
```

**Comment by benaadams** [comment]
If its a `Timer` being kicked off that then runs Sql queries you will likely want to suspend `ExecutionContext` flow when creating that `Timer` so that the `HttpContext` doesn't get captured and become forever accessible via `IHttpContextAccessor` when it shouldn't be (e.g. its only via during that request)

Using a pattern similar to https://github.com/dotnet/corefx/pull/26075

**Comment by benaadams** [comment]
Or aspnet's NonCapturingTimer helper https://github.com/aspnet/Common/pull/395

**Comment by Svisstack** [comment]
@benaadams Thanks, I actually found the Tasks which can and probably are ending after the requests is finished with the SQL code, I will wait on them and report about the state of the issue.

Based on your very useful input, I think it's still issued in the AspNetCore as it should not provide functions which delivering possible random data, if Headers dictionary is shared between requests then function for accessing that headers should have required some kind of request ID (GUID) based on which internal function could decide that these headers no longer exists.

**Comment by benaadams** [comment]
>  I think it's still issue in the AspNetCore as it should not provide functions which delivering possible random data, if Headers dictionary is shared between requests

Its more the Headers for the connection as their can only be one active request on a connection at any time. If they are accessed via `IHttpContextAccessor` there is a change in ASP.NET Core 2.2 to invalidate the HttpContext and return null in the manner you described https://github.com/aspnet/HttpAbstractions/pull/1021

However, while that may help with the CPU issue; it would also mean the telemetry from `OnStart` would be disconnected from `OnEnd`; if its using anything in the headers or HttpContext to connect them (as `OnEndXxx` wouldn't have this information).

What should probably happen is Application Insights during `OnStartXxx`; which happens during the request, should deep copy the information it needs from the context/headers (e.g. key, values) as the strings will remain valid being immutable. Then `OnEndXxx` should use that same copied object to get the headers which were captured at the start of the event.

I'm not entirely sure how Application Insights is capturing/reading the headers so I can't provide code suggestions for the changes, alas...

**Comment by Svisstack** [comment]
@benaadams Removing the Tasks with SQL code that could end after the request solved the issue. Many thanks, I really appreciate your input here. 

So it's no action needed here as there is a new interface which didn't have this issue? I want to make sure that we should fix that for the community, or at least in known Insights bugs should be information that no dependency should end after the request is already finished.

**Comment by benaadams** [comment]
Well I do think AI should collect the information it needs on event start (i.e. key,value for headers); then use that collected information for event end; rather than rereading it from the HttpHeader collection; which should solve this issue more holistically.

**Comment by Svisstack** [comment]
@benaadams Can you then reopen this issue?

**Comment by benaadams** [comment]
>  Can you then reopen this issue?

Nope, I don't work for MS or have rights to this repo

**Comment by cijothomas** [comment]
@Svisstack Can you open this issue in https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues? 
From an initial look, the issue you are facing is from DependencyCollection module, which is in not in this repo. (This repo only has asp.net core extension methods, and request collection)
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues

**Comment by Svisstack** [comment]
@cijothomas It's done.

**Comment by cijothomas** [comment]
@Svisstack thanks. Will discuss on othed thread.

**Comment by turgayozgur** [comment]
> @benaadams Thanks, I actually found the Tasks which can and probably are ending after the request is finished and they contain the SQL code, I will wait on them and report about the state of the issue after deploying the new version today.
> 
> Based on your very useful input, I think it's still issue in the AspNetCore as it should not provide functions which delivering possible random data, if Headers dictionary is shared between requests then function for accessing that headers should have required some kind of request ID (static long incrementing every new request) based on which internal function could decide that these headers no longer exists and throw an exception instead of providing random data.

Hi @Svisstack,

How did you solve that with this way? Could you share any code detail?

**Comment by Svisstack** [comment]
@Turgay I was able to wait on all the tasks or rewriting a sync code into the sync one before the request ended, in the future I could not probably do that then I believe migrating to the other lib or waiting for the fix are the only options.

**Comment by turgay** [comment]
@Svisstack I guess you meant @turgayozgur but not me :)
